### PR TITLE
fix: make `firebase-tools` a local dependency where needed

### LINKED ISSUE #771: Firebase config variables are not available when deploying functions
#### Bug, feature request, or proposal:

I'm experiencing a bug in which I can't deploy the firebase functions due to firebase config variables not being available. This might be a problem with unsupported package versions. See below for more information.

#### What is the expected behavior?

```
$ npm run deploy
```

From within `firebase/functions` should run successfully

#### What is the current behavior?

When running the deploy script, firebase errors out with the error:

```
Error: Error occurred while parsing your function triggers. Please ensure you have the latest firebase-functions SDK by running "npm i --save firebase-functions@latest" inside your functions folder.

Error: Firebase config variables are not available. Please use the latest version of the Firebase CLI to deploy this function.
    at init (/Users/pascalprecht/projects/machinelabs/machinelabs/firebase/functions/dist/node_modules/firebase-functions/lib/config.js:51:15)
    at Object.config (/Users/pascalprecht/projects/machinelabs/machinelabs/firebase/functions/dist/node_modules/firebase-functions/lib/config.js:29:9)
    at Object.<anonymous> (/Users/pascalprecht/projects/machinelabs/machinelabs/firebase/functions/dist/invocation-processor/database.js:5:42)
    at Module._compile (module.js:662:30)
    at Object.Module._extensions..js (module.js:673:10)
    at Module.load (module.js:575:32)
    at tryModuleLoad (module.js:515:12)
    at Function.Module._load (module.js:507:3)
    at Module.require (module.js:606:17)
    at require (internal/module.js:11:18)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @machinelabs/firebase-functions@0.28.1 deploy: `npm run build && firebase deploy && npm run clean_dist`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @machinelabs/firebase-functions@0.28.1 deploy script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/pascalprecht/.npm/_logs/2018-03-24T10_04_32_198Z-debug.log
```

This is very likely caused by those lines of code:

```
const serviceAccount = functions.config().fb_service_account;

if (serviceAccount) {
  conf.credential = admin.credential.cert(<any>{
    private_key: serviceAccount.private_key,
    client_email: serviceAccount.client_email
  });
}
```

So I went ahead and double checked my config variables with

```
$ firebase functions:config:get
```

which outputs my configured `client_email` and `private_key`. Both are correct.

#### Which version of MachineLabs (or the package) is affected ?

`@machinelabs/firebase-functions@27d92f19`

#### Is there anything else we should know?

I digged a bit further and ran into https://github.com/firebase/firebase-tools/issues/609#issuecomment-359042237, which suggests that `firebase-tools` and `firebase-functions` SDK need to be both on latest. 

Which makes me think a possible reason for this issue is that I haven't done any firebase deployments with my current setup so far (new laptop, and we didn't introduce any changes to functions/rules etc), however, I'm running on `firebase-tools@3.17.7` (globally), which might not be compatible anymore with `firebase-functions@0.5.9`. That's the version that currently gets installed.

I will have to try older versions of `firebase-tools` to verify whether this is causing this or not.

**Sidenote**: On that note, we might want to consider switching to `npx` for firebase-tools as well?

Anyways, would be great if anyone can try it out by installing the latest firebase-tools and verifying whether this is reproducable or not.

### PR DESCRIPTION
**Pull request by PascalPrecht**
This also changes npm scripts to use `npx firebase` to ensure the
local version is used.

Fixes #771

### CODE REVIEW

**Comment by PascalPrecht** [comment]
@cburgdorf would be cool if you could test admin-cli deployment in this version

**Comment by d3lm** [comment]
The code looks good, but haven't tested it locally yet.

**Comment by PascalPrecht** [comment]
@d3lm Thanks for the review! Found some time to test it yet?

Also pinging @cburgdorf again to give the admin-cli deployment a try on staging.

**Comment by cburgdorf** [comment]
Ok, I gave this a test as I tried to make a deployment on staging. Everything went smooth at first but when it arrived at the stage where the client is deployed it failed.

Here is what happened:

1. My node version was out of date. This is probably unrelated to this PR and more related to the PR that introduced prettier. No problem so I updated my node version but now I'm running into the following.

2. I get a `/bin/sh: 3: ng: not found`. This is because I haven't installed the angular-cli globally yet after I updated `node`. However, with `npx` am I right to assume that a global `angular-cli` is actually not what we want?

**Comment by PascalPrecht** [comment]
> I get a /bin/sh: 3: ng: not found. This is because I haven't installed the angular-cli globally yet after I updated node. However, with npx am I right to assume that a global angular-cli is actually not what we want?

The only need for a global `ng` is convenience and scaffolding of projects. Whenever we use any `ng` command inside a CLI project, what actually happens is that Angular CLI will use the local one all the time.

So what I think we **could** do is running `ng` through `npx` as well. Since this the version that's used anyways, it's jut not accessible via a global `ng` command then.

**Comment by PascalPrecht** [comment]
@cburgdorf updated the PR accordingly. Maybe you can give it another shot.

**Comment by cburgdorf** [comment]
Tried it! Works now. We can merge it as soon as it goes green.

File: firebase/functions/package.json [hunk 1]
```diff
@@ -9,7 +9,7 @@
     "node_modules": "rm -rf ./node_modules && npx yarn install",
     "build:pre": "npm run node_modules && npm run clean_dist && npm run mk_dist && npm run copy",
     "build": "npm run build:pre && tsc",
-    "deploy": "npm run build && firebase deploy && npm run clean_dist",
+    "deploy": "npm run build && npx firebase deploy && npm run clean_dist",
```
Comment (SamVerschueren) [inline firebase/functions/package.json]: This should automatically pick the one from `node_modules`. So it should also work without `npx`.

File: rest-api/package.json [hunk 2]
```diff
@@ -23,7 +23,7 @@
     "moduleFileExtensions": ["ts", "tsx", "js", "json"]
   },
   "scripts": {
-    "node_modules": "rm -rf ./node_modules && yarn install",
+    "node_modules": "rm -rf ./node_modules && npx yarn install",
```
Comment (SamVerschueren) [inline rest-api/package.json]: Same as above.

**Comment by SamVerschueren** [comment]
Just added a small comment, feel free to ignore me :).

**Comment by PascalPrecht** [comment]
Thanks @SamVerschueren ! Yes those are left overs and can be removed. I'll take care of it!

**Comment by PascalPrecht** [comment]
Okay updated the PR. Will merge this once green on Travis.

**Pull request merged**